### PR TITLE
include version requirements in path for resolver error messages

### PR DESCRIPTION
**Pull request by Eh2406**
This is a follow up to https://github.com/rust-lang/cargo/pull/5428#issuecomment-385477365. It changes the error messages from:

https://github.com/rust-lang/cargo/blob/0b530c30867da26a4b59146f490c9f1d5377c20a/tests/testsuite/build.rs#L1356-L1358

to:

https://github.com/rust-lang/cargo/blob/499340a81624eaef77d969a3e9fd2cf94f933a2e/tests/testsuite/build.rs#L1356-L1358

Sorry for the big diff, this includes a `cargo +stable fmt` commit before my changes.

### CODE REVIEW

**Comment by alexcrichton** [comment]
This looks pretty awesome, a great idea!

I wonder though if we may want to tweak the wording slightly? It may be slightly redundant to have the name of the package like `foo = "0.1"` because it's repeated on every line, and I think it's also not necessarily guaranteed to match what's written in the manifest as well. 

It may also be worth avoiding printing a `*` dependency requirement for things like git/path dependencies, but I'm not acutally sure how that could cause a conflict so it may not be too relevant!

**Comment by Eh2406** [comment]
Rebase took a couple of tries, but doing it manually worked.

I removed the redundant use of `name`. Do you have a less verbose/stilted way of saying this? And, a preferred wording for git/path?

**Comment by matklad** [comment]
I definitely like the information that is presented, however all these `... selected to fulfill` are hard to parse for me. I would say that there are two problems with this presentation: 

* package and it's version requirement are on adjacent lines, so you have to read this stuff diagonally to make sense
* we try hard to present a tabular data as a single English sentence, and that feels clunky.

As alternative, what about this more line oriented and tabular presentation:

```
BEFORE:
error: failed to select a version for `bad`.
    ... required by package `incompatible_dependencies v0.0.1 ([..])`
versions that meet the requirements `>= 1.0.1, <= 2.0.0` are: 2.0.0, 1.0.1
all possible versions conflict with previously selected packages.
  previously selected package `bad v2.0.1`
    ... selected to fulfill the requirement \">= 2.0.1\" from package `baz v0.1.0`
    ... selected to fulfill the requirement \"^0.1.0\" from package `incompatible_dependencies v0.0.1 ([..])`
  previously selected package `bad v1.0.0`
    ... selected to fulfill the requirement \"= 1.0.0\" from package `bar v0.1.0`
    ... selected to fulfill the requirement \"^0.1.0\" from package `incompatible_dependencies v0.0.1 ([..])`
failed to select a version for `bad` which could resolve this conflict

AFTER:
error: failed to select a version for `bad`.
   `bad >= 1.0.1, <= 2.0.0` required by `incompatible_dependencies v0.0.1 ([..])`

versions that meet the requirements `bad >= 1.0.1, <= 2.0.0` are: 2.0.0, 1.0.1
all possible versions conflict with previously selected packages.

previously selected package `bad v2.0.1`
  `bad >= 2.0.1` required by `baz v0.1.0`, picked  `bad v2.0.1`
  `baz ^0.1.0` required by `incompatible_dependencies v0.0.1 ([..])`, picked `baz v0.1.0`

previously selected package `bad v1.0.0`
  `bad  =1.0.0` required by `bar v0.1.0`, picked  `bad v1.0.0`
  `bar ^0.1.0` required by `incompatible_dependencies v0.0.1 ([..])`,  picked `bar v0.1.0`

failed to select a version for `bad` which could resolve this conflict",
```

I am not really attached to this format, but I like some of its properties:

* it purposefully duplicates version info, so as to make each line complete in isolation
* version requirements (which are the underling cause of the conflict) are in the first column. 
* simpler and shorter English words 
* blank lines :heart: :heart: :heart:

**Comment by matklad** [comment]
As this is hard to rebase, I suggest to merge this ASAP, and tweak rendering details in subsequent PR though, if we want to tweak them (I actually quite like the current rendering, I just think it could be even better).

**Comment by matklad** [comment]
>Rebase took a couple of tries, but doing it manually worked.

FWIW, I always struggle with rebasing if I send PR from the master branch (and this is what happened here I guess?). My current setup is to always send PR from dedicated branches, and have a `master` branch of my local checkout of Cargo set to track `master` branch from `rust-lang/cargo` and not `matklad/cargo`.

**Comment by Eh2406** [comment]
as pointed out in #5484, we should mention that precise requirements may be caused by lock files and suggest using update or generate lock file to fix it.

**Comment by alexcrichton** [comment]
Oh dear it looks like this may have fallen off the radar by accident! @Eh2406 was there any other blockers on this or did you feel this was good to go?

**Comment by Eh2406** [comment]
It fell off my radar as well. What was left to do:
1. Agree on desired wording.
2. Special case git/path dependencies without breaking the structure of the output.
3. Special case pind by a lock file and not a `"=1.1.2"` in a dependency. Also add a "note: try cargo update" to the end.

**Comment by alexcrichton** [comment]
Ah ok, no worries!

**Comment by alexcrichton** [comment]
Ok I'm gonna close this for now due to inactivity, but we can of course reopen and/or resubmit!
